scripts/convert-to-optional.js

```diff
@@ -0,0 +1,252 @@
+const babel = require("@babel/core");
+const fs = require("fs");
+const path = require("path");
+
+/**
+ * Reference on AST types:
+ * - Docs on babel types https://babeljs.io/docs/babel-types
+ * - Discover and play with AST https://astexplorer.net/ (select @babel/parser as parser)
+ */
+
+/**
+ * Identifies the `[] | [T]` pattern
+ */
+const isOptionalUnionType = (annotation) => {
+  if (annotation.typeAnnotation.type === "TSUnionType") {
+    const types = annotation.typeAnnotation.types;
+    const hasEmptyTuple = types.some(
+      (type) => type.type === "TSTupleType" && type.elementTypes.length === 0
+    );
+    const hasOneElementTyple = types.some(
+      (type) => type.type === "TSTupleType" && type.elementTypes.length === 1
+    );
+    return types.length === 2 && hasEmptyTuple && hasOneElementTyple;
+  }
+};
+
+const EXTERNAL_TYPES = ["Principal", "Uint8Array"];
+
+const mappedInterfaces = new Map();
+
+const transformOptionalTuple = () => {
+  return {
+    visitor: {
+      TSInterfaceDeclaration: (path) => {
+        const name = path.node.id.name;
+        if (name.startsWith("Optional") || name.startsWith("_")) {
+          return;
+        }
+        const t = babel.types;
+        const convertTypeToOptional = (node) => {
+          if (node.elementTypes.length === 0) {
+            return t.tsUndefinedKeyword();
+          }
+          if (node.elementTypes.length === 1) {
+            return node.elementTypes[0];
+          }
+          throw new Error(`Unexpected tuple type ${JSON.stringify(node)}`);
+        };
+        const convertNode = (node) => {
+          const annotation = node.typeAnnotation;
+          if (isOptionalUnionType(annotation)) {
+            return t.tsPropertySignature(
+              node.key,
+              t.tsTypeAnnotation(
+                t.tsUnionType(
+                  annotation.typeAnnotation.types.map(convertTypeToOptional)
+                )
+              )
+            );
+          }
+          return node;
+        };
+        const interfaceName = path.node.id.name;
+        const newInterfaceName = `Optional${interfaceName}`;
+        const newIdentitifer = t.identifier(newInterfaceName);
+        const newBody = path.node.body.body.map(convertNode);
+        const newInterface = t.tsInterfaceDeclaration(
+          newIdentitifer,
+          null,
+          [],
+          t.tSInterfaceBody(newBody)
+        );
+
+        const parameterName =
+          interfaceName.charAt(0).toLowerCase() + interfaceName.slice(1);
+        const parameter = t.identifier(parameterName);
+        parameter.typeAnnotation = t.tsTypeAnnotation(
+          t.tsUnionType([
+            t.tsTypeReference(t.identifier(interfaceName)),
+            t.tsUndefinedKeyword(),
+          ])
+        );
+        const transformProp = (node) => {
```

Comment:
Probably having a comment w/ a sample argument structure could help with reading the code.